src/main/java/edu/harvard/iq/dataverse/engine/command/impl/PublishDatasetCommand.java

```diff
@@ -75,7 +74,17 @@ public PublishDatasetResult execute(CommandContext ctxt) throws CommandException
         boolean reservingPidsSupported = !idServiceBean.registerWhenPublished();
         if (reservingPidsSupported) {
             if (theDataset.getGlobalIdCreateTime() == null) {
-                throw new IllegalCommandException(BundleUtil.getStringFromBundle("publishDatasetCommand.pidNotReserved"), this);
+            	//Backward compatibility and recovery from PID provider outage - if we should have reserved at creation time but didn't, try here as well to keep moving. If we still can't reserve (service down, connflict, etc.) then bail
+            	try {
+                    String returnString = idServiceBean.createIdentifier(this.getDataset());
```

Comment:
If we are going back to trying to register an unreserved DOI during publishing, it should probably happen in the other, asynchronous part of the workflow - in FinalizeDatasetPublicationCommand. Otherwise we can again end up in a situation where Datacite decides to hang for a while, and this is happening while the dataset is still unlocked, etc.
And in the finalize command we already have this code:
```
if ( theDataset.getGlobalIdCreateTime() == null ) {
   try {
      registerExternalIdentifier(theDataset, ctxt);
   }
   ...
```
but of course that registerExternalIdentifier method will need to be modified, because in its current form it would get a new pid assigned to the dataset, if the old one already exists on the Datacite end...

Reply:
Does DataCite hang/delay an http call (versus fail)? If so I can move this/integrate with registerExternalIdentifier to avoid waiting for the call to succeed/fail. It seems like going async and not being able to respond to the user when the first step fails (before file validation, file DOI registrations) except with the async notification isn't great either, but if there are cases where reserving the dataset DOI might induce people to click again, I guess it makes sense.

Reply:
Well, it is a network call - so it can hang... I believe we have seen registration calls actually taking a long time and/or timing out, during the worst of their service troubles.
But yes, one nice thing about the current implementation is that we can exit right away with a clear message, before the dataset gets locked and the command goes async.
(we could alternatively try and leave it in the first part of the command - but make the call with a forced short timeout?)

Reply:
OK - updated registerExternalIdentifier and moved things to Finalize. As far as I can tell, the httpClient should be limiting the wait for a response to one minute (system default) but that could be lowered in the DataCIteRESTfullClient (easiest to set once for all calls since there's one shared client) and I think the whole call to registerExternalIdentifier could move to Publish then with a cut/paste.